cpp/src/arrow/adapters/orc/adapter_util.cc

```diff
@@ -315,13 +344,662 @@ Status AppendBatch(const liborc::Type* type, liborc::ColumnVectorBatch* batch,
+      const uint64_t precision = static_cast<uint64_t>(
+          static_cast<const arrow::Decimal128Type&>(type).precision());
+      const uint64_t scale =
+          static_cast<uint64_t>(static_cast<const arrow::Decimal128Type&>(type).scale());
+      return liborc::createDecimalType(precision, scale);
+    }
+    case arrow::Type::type::LIST:
+    case arrow::Type::type::FIXED_SIZE_LIST:
+    case arrow::Type::type::LARGE_LIST: {
+      std::shared_ptr<arrow::DataType> arrow_child_type =
+          static_cast<const arrow::BaseListType&>(type).value_type();
+      ORC_UNIQUE_PTR<liborc::Type> orc_subtype =
+          GetORCType(*arrow_child_type).ValueOrDie();
+      return liborc::createListType(std::move(orc_subtype));
+    }
+    case arrow::Type::type::STRUCT: {
+      ORC_UNIQUE_PTR<liborc::Type> out_type = liborc::createStructType();
+      std::vector<std::shared_ptr<arrow::Field>> arrow_fields =
+          checked_cast<const arrow::StructType&>(type).fields();
+      for (std::vector<std::shared_ptr<arrow::Field>>::iterator it = arrow_fields.begin();
+           it != arrow_fields.end(); ++it) {
+        std::string field_name = (*it)->name();
+        std::shared_ptr<arrow::DataType> arrow_child_type = (*it)->type();
+        ORC_UNIQUE_PTR<liborc::Type> orc_subtype =
+            GetORCType(*arrow_child_type).ValueOrDie();
+        out_type->addStructField(field_name, std::move(orc_subtype));
+      }
+      return out_type;
+    }
+    case arrow::Type::type::MAP: {
+      std::shared_ptr<arrow::DataType> key_arrow_type =
+          checked_cast<const arrow::MapType&>(type).key_type();
+      std::shared_ptr<arrow::DataType> item_arrow_type =
+          checked_cast<const arrow::MapType&>(type).item_type();
+      ORC_UNIQUE_PTR<liborc::Type> key_orc_type =
+                                       GetORCType(*key_arrow_type).ValueOrDie(),
+                                   item_orc_type =
+                                       GetORCType(*item_arrow_type).ValueOrDie();
+      return liborc::createMapType(std::move(key_orc_type), std::move(item_orc_type));
+    }
+    case arrow::Type::type::DENSE_UNION:
+    case arrow::Type::type::SPARSE_UNION: {
+      ORC_UNIQUE_PTR<liborc::Type> out_type = liborc::createUnionType();
+      std::vector<std::shared_ptr<arrow::Field>> arrow_fields =
+          checked_cast<const arrow::UnionType&>(type).fields();
+      for (std::vector<std::shared_ptr<arrow::Field>>::iterator it = arrow_fields.begin();
+           it != arrow_fields.end(); ++it) {
+        std::string field_name = (*it)->name();
+        std::shared_ptr<arrow::DataType> arrow_child_type = (*it)->type();
+        ORC_UNIQUE_PTR<liborc::Type> orc_subtype =
+            GetORCType(*arrow_child_type).ValueOrDie();
+        out_type->addUnionChild(std::move(orc_subtype));
+      }
+      return out_type;
+    }
+    default: {
+      return arrow::Status::Invalid("Unknown or unsupported Arrow type: ",
```

Comment:
Here as well, either `Status::TypeError` or `Status::NotImplemented`

Reply:
Thanks! Switched to `Status::NotImplemented` since we don't have dictionary support in this PR but it will be supported.